### DISCUSSION TITLE
Find/Query all photos with no geotag

### DISCUSSION THREAD

**Discussion by daniambrosio** [RhetTbull/osxphotos#468] [Q&A]
Hey guys.... I was looking for a way to query all the photos that do not have a geotag or GPS location information, but could find a way to do this.

Am I missing something or is this functionality really not available?

**Comment by RhetTbull** [comment]
@daniambrosio I'm not at my Mac so can't test this but I think this will work with `osxphotos query` or `osxphotos export`:

`--query-eval "photo.location == (None, None)"`

`--query-eval` evaluates a python statement to filter photos.  `photo.location` returns a tuple (pair) of `(latitude, longitude)` and these will be set to the special value `None` if there is no location specified in the photo.

**Reply by daniambrosio** [reply]
Thanks for your reply... just found this one:

"--no-place                   Search for photos with no associated place name info (no reverse geolocation info)"

Will try both and come back here!

**Reply by RhetTbull** [reply]
`--no-place` is slightly different than "no GPS location information".  The --place and --no-place look at the reverse geolocation information (that is, the name of the place such as "Los Angeles, California, United States" ) that Apple Photos adds to photos automatically.  It is possible that you could have a photo that has GPS information associated with it but does not have the reverse geolocation info (place).  For example, if Apple hadn't yet indexed it or if no reverse geolocation info was found for that particular latitude and longitude.

**Reply by daniambrosio** [reply]
Thanks! Super clear and accurate response according to the tests I was doing. 
However the result does not match to a Smart Album using a "no GPS" filter.

**Reply by RhetTbull** [reply]
Interesting.  How was osxphotos result different than the Smart Album?  Did osxphotos show more photos or did the Smart Album?  Where the different by a lot or just a few photos?  I'm not sure what fields the Smart Album is looking at -- will do some testing to see if I can figure it out.  Osxphotos is looking at the location data that is either imported from the original EXIF or added by the user in Photos.

**Reply by RhetTbull** [reply]
Just did some more testing.  Is this an iCloud library (e.g. your photos are synched to iCloud) and if so, do you have photos in shared albums?  If so, try adding `--not-shared` to your osxphotos options.  By default, osxphotos includes all photos including those that are in shared albums.  Photos does not do this.  So the photos in the Smart Album do not include any shared albums.  When I did the following:

`osxphotos query --query-eval "photo.location == (None, None)" --not-shared  > nogps.txt`

the results matched the expected number of photos in the No GPS Smart Album.

**Comment by daniambrosio** [comment]
Been some days away... I can come back to you in a few days! Thanks for
looking into it!

On Tue, Jun 15, 2021 at 3:21 AM Rhet Turnbull ***@***.***>
wrote:

> Just did some more testing. Is this an iCloud library (e.g. your photos
> are synched to iCloud) and if so, do you have photos in shared albums? If
> so, try adding --not-shared to your osxphotos options. By default,
> osxphotos includes all photos including those that are in shared albums.
> Photos does not do this. So the photos in the Smart Album do not include
> any shared albums. When I did the following:
>
> osxphotos query --query-eval "photo.location == (None, None)" --not-shared
> > nogps.txt
>
> the results matched the expected number of photos in the No GPS Smart
> Album.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RhetTbull/osxphotos/discussions/468#discussioncomment-871605>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAHFFQYLDOSGAKAJWQG3PTTS3WM3ANCNFSM46VZLHTQ>
> .
>


-- 
-------------
Daniel R. Ambrosio
***@***.***

**Comment by RhetTbull** [comment]
I've added `--location` and `--no-location` in [v0.42.44](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.44) to find photos that contain location information or do not contain location information.  Unlike `--place`, `--no-place` these look for photos that contain latitude/longitude coordinates not the reverse geolocation place names.

**Comment by daniambrosio** [comment]
Cool... will try this new flag. 
As promised I bring you the info on how different the results are. 

- I am using a Smart Album with the filter "Photo / is not / GPS info" - this results in 676 total

The following are the results using v0.42.34
- Running the command "osxphotos query --json --query-eval "photo.location == (None, None)"" - this results in 1110 total
- Adding the "--not-shared" flag to the above - this results in 676 total

- Running the command "osxphotos query --json --no-place" - this results in 1776 total
- Adding the "--not-shared" flag to the above - this results in 902 total

- Lastly, after upgrading to v0.42.45 I came with the following result when running "osxphotos query --json --no-location" - this results in 1110 total (just like --query-eval "photo.location == (None, None)")

**Reply by RhetTbull** [reply]
Thanks!  Looks like excluding shared images provides the same result as Photos (as expected).
